pkg/querier/lazy_querier.go

```
@@ -75,28 +85,43 @@ func (e errSeriesSet) Err() error {
 }
 
 type lazySeriesSet struct {
+	ctx    context.Context
 	next   storage.SeriesSet
 	future chan storage.SeriesSet
 }
 
+// Waits for series set to be ready and returns true, if s.next has been set and can be used.
+func (s *lazySeriesSet) waitForSeriesSet() bool {
+	if s.next != nil {
+		return true
+	}
+
+	select {
+	case s.next = <-s.future:
+		return s.next != nil
+	case <-s.ctx.Done():
+		return false
+	}
+}
+
 // Next implements storage.SeriesSet.  NB not thread safe!
 func (s *lazySeriesSet) Next() bool {
-	if s.next == nil {
-		s.next = <-s.future
+	if !s.waitForSeriesSet() {
+		return false
 	}
 	return s.next.Next()
 }
 
 func (s lazySeriesSet) At() storage.Series {
-	if s.next == nil {
-		s.next = <-s.future
+	if !s.waitForSeriesSet() {
+		return nil
 	}
 	return s.next.At()
 }
 
 func (s lazySeriesSet) Err() error {
-	if s.next == nil {
-		s.next = <-s.future
+	if !s.waitForSeriesSet() {
+		return nil
```

Is it correct return `nil` here? Shouldn't we return the context error?